### PR TITLE
Separate input_ids and decoder_input_ids in model.generate()

### PR DESCRIPTION
**Pull request by yjernite**
This makes the generation behavior more similar for sequence-to-sequence and language models, and allows us to initialize decoding with a prefix for the encoder-decoder setting.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -800,6 +821,19 @@ def generate(
                 "Please use another model class (e.g. `OpenAIGPTLMHeadModel`, `XLNetLMHeadModel`, `GPT2LMHeadModel`, `CTRLLMHeadModel`, `T5WithLMHeadModel`, `TransfoXLLMHeadModel`, `XLMWithLMHeadModel`, `BartForConditionalGeneration` )"
             )
 
+        if self.config.is_encoder_decoder:
+            bos_token_id = decoder_start_token_id if decoder_start_token_id is not None else bos_token_id
+            bos_token_id = bos_token_id if bos_token_id is not None else self.config.decoder_start_token_id
+            bos_token_id = bos_token_id if bos_token_id is not None else self.config.bos_token_id
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:827]: How did you come up with this hierarchy?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -658,6 +658,11 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
     def generate(
         self,
         input_ids=None,
+        encoder_input_ids=None,
+        decoder_input_ids=None,
+        attention_mask=None,
+        encoder_attention_mask=None,
+        decoder_attention_mask=None,
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:665]: 1) what if attention_mask and enc_attention_mask both specified?
2) when would you need decoder_attention_mask?

Comment (sshleifer) [reply src/transformers/modeling_utils.py:665]: The naming is confusing because this decoder_attention_mask does not correspond to argument names for seq2seq models. I had to read the code to figure out what was going on.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -885,10 +923,15 @@ def generate(
...
         # TODO (PVP): this should later be handled by the forward fn() in each model in the future see PR 3140
-        if (attention_mask is None) and (pad_token_id is not None) and (pad_token_id in input_ids):
-            attention_mask = input_ids.ne(pad_token_id).long()
-        elif attention_mask is None:
-            attention_mask = input_ids.new_ones(input_ids.shape)
+        if (decoder_attention_mask is None) and (pad_token_id is not None) and (pad_token_id in decoder_input_ids):
+            decoder_attention_mask = decoder_input_ids.ne(pad_token_id).long()
+        elif decoder_attention_mask is None:
+            decoder_attention_mask = decoder_input_ids.new_ones(decoder_input_ids.shape)
+        if self.config.is_encoder_decoder:
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:930]: Im fine with making masks on the fly, but other hf apis dont do this (they let batch_encode_plus do it )

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -885,10 +923,15 @@ def generate(
...
+            if (encoder_attention_mask is None) and (pad_token_id is not None) and (pad_token_id in encoder_input_ids):
+                encoder_attention_mask = encoder_input_ids.ne(pad_token_id).long()
+            elif encoder_attention_mask is None:
+                encoder_attention_mask = encoder_input_ids.new_ones(encoder_input_ids.shape)
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:934]: If you're going to make the mask anyway, the `(pad_token_id in encoder_input_ids)` check is unnecessary.

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -885,10 +923,15 @@ def generate(
...
-        if (attention_mask is None) and (pad_token_id is not None) and (pad_token_id in input_ids):
-            attention_mask = input_ids.ne(pad_token_id).long()
-        elif attention_mask is None:
-            attention_mask = input_ids.new_ones(input_ids.shape)
+        if (decoder_attention_mask is None) and (pad_token_id is not None) and (pad_token_id in decoder_input_ids):
+            decoder_attention_mask = decoder_input_ids.ne(pad_token_id).long()
+        elif decoder_attention_mask is None:
+            decoder_attention_mask = decoder_input_ids.new_ones(decoder_input_ids.shape)
+        if self.config.is_encoder_decoder:
+            if (encoder_attention_mask is None) and (pad_token_id is not None) and (pad_token_id in encoder_input_ids):
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py:931]: helper_fn?

**Review by sshleifer — COMMENTED** [review]
Confused by all the ids floating around and turning into each other.

Also, would love tests.

**Comment by sshleifer** [comment]
Can we get the prefix functionality with just one kwarg instead of all the renaming?

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -800,6 +821,19 @@ def generate(
...
             )
 
+        if self.config.is_encoder_decoder:
+            bos_token_id = decoder_start_token_id if decoder_start_token_id is not None else bos_token_id
+            bos_token_id = bos_token_id if bos_token_id is not None else self.config.decoder_start_token_id
+            bos_token_id = bos_token_id if bos_token_id is not None else self.config.bos_token_id
+            assert (
+                bos_token_id is not None
+            ), "decoder_start_token_id or bos_token_id has to be defined for encoder-decoder generation"
+            encoder_input_ids = encoder_input_ids if encoder_input_ids is not None else input_ids
```
[the remark this answers is not in the document]
Comment (yjernite) [reply src/transformers/modeling_utils.py:831]: I really think the main arguments should have the same behavior whether we're using generate in the language model or seq2seq model: so encoder_input_ids goes to the encoder, and decoder_input_ids goes to the decoder (a language model behaves broadly like a decoder). I kept input_ids as an alias so as not to break any of the current examples, but it should probably be deprecated. If someone passes all three, input_ids will be ignored.

In the previous seq2seq setting, input_ids was first used as input to the encoder then completely overwritten to hold the input of the decoder, which I definitely had some trouble wrapping my head around.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (yjernite) [reply src/transformers/modeling_utils.py:827]: Trying not to break the current examples, but I'm also not sure why we should have both decoder_start_token_id and bos_token_id. I don't think they were used together anywhere in the previous code.

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (yjernite) [reply src/transformers/modeling_utils.py:665]: I was confused about the uses of attention-mask, hadn't realized that seq2seq models only take encoder-side attention. Will switch that back to a single attention_mask value in the next commit.

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (yjernite) [reply src/transformers/modeling_utils.py:930]: That part I copied over from the previous version of the code, but I think that @patrickvonplaten had a TODO to remove it anyways

**Comment by yjernite** [comment]
Got stuck in a merge / rebase loop, closing and starting again.
